### PR TITLE
New branch

### PR DESCRIPTION
**Pull request by ajithnn**

### CODE REVIEW

File: mygit.rb [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+require 'optparse/time'
+require 'octokit'
+require 'io/console'
+require 'logger'
+
+$force = 'False'
+
+options ={
+:files => [],
+:repoandbranch => [],
```
Comment (ajithnn) [inline mygit.rb:11]: use underscore case. Like below
repo_and_branch

File: mygit.rb [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+:force => 'False',
+:reviewers => []
+}
+
+OptionParser.new do |parser|
+   #Handling the -create Command
+
+   parser.on('-c', '--create=repo branch',Array, 'Create a Repo and Branch') do |create|
+        options[:repoandbranch] += create  
+        mygit_create(options[:repoandbranch])
```
Comment (ajithnn) [inline mygit.rb:21]: function name can be just create , mygit is unnecessary

File: mygit.rb [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+   #Handling the Check Status --status command
+
+   parser.on('-s','--status', 'Check Status of recent pull request and review request') do |status|
+        mygit_status()
+   end
+
+   #Handling the Git Configuration 
+
+   parser.on('--config', 'Configure Git') do
+        configExist?()
```
Comment (ajithnn) [inline mygit.rb:33]: casing should be consistent , either camel or snake.
In ruby the convention is snake case. so config_exists?

File: mygit.rb [hunk 4]
```diff
@@ -0,0 +1,365 @@
+require 'optparse'
+require 'optparse/time'
+require 'octokit'
+require 'io/console'
+require 'logger'
+
+$force = 'False'
```
Comment (ajithnn) [inline mygit.rb:7]: If the option parsing and handling functions were to be inside a class , then global variables will not be necessary.

File: mygit.rb [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+   parser.on('-f', '--force', 'Force Push') do |force|
+        $force='True'
+   end
+
+
+options[:files] += ARGV
+options[:repoandbranch] += ARGV
+options[:reviewers] += ARGV
+
+$LOG = Logger.new('log_file.log', 'monthly')
```
Comment (ajithnn) [inline mygit.rb:63]: Pass log file path as an option

File: mygit.rb [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+
+def configExist?()
```
Comment (ajithnn) [inline mygit.rb:65]: everything to snake_case

File: mygit.rb [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+                if ARGV.include?('-f')
+                        $LOG.debug("<MyGit> : <ConfigExist?> : Forcefully creating a new config")
+                        ConfigGit()
+                        return                        
+                end
+                puts "<MyGit> : Config Exists already \n use -f tag to force a new config"
+                $LOG.debug("<MyGit> : <ConfigExist?> : Config Already Exists!")
+         else
+                $LOG.debug("<MyGit> : <ConfigExist?> : Calling the ConfigGit function to configure the user's git")
+                Dir.chdir('../..')
```
Comment (ajithnn) [inline mygit.rb:81]: Why chdir ?

File: mygit.rb [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+        else
+                $LOG.debug("<MyGit> : <GitExist?> : Git isnt installed, requesting the user to install git")
+                puts "<MyGit> : <Error> : git isn`t installed on your system, kindly install and run the app again!"
+                return false
+        end
+        
+end     
+
+
+def getreponame
```
Comment (ajithnn) [inline mygit.rb:145]: Again naming is inconsistent : get_repo_name

File: mygit.rb [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+        $LOG.debug("<MyGit> : <getreponame> : Asking for reponame")
+        repo = File.open('repo.txt', 'r').read
```
Comment (ajithnn) [inline mygit.rb:147]: File is not closed , also reusing file variable to assign string is bad practice.

For line by line reading of file 
```ruby
File.open("path","r").each do |line|
end
```

each_with_index also exists if you need only the second line .

File: mygit.rb [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+        repo = repo.split('/')[1]
+        return repo
+
+end
+
+def mygit_create(repoandbranch)
+        $LOG.debug("<MyGit> <mygit_create> : mygit_create called with repo: #{repoandbranch[1]} and branch: #{repoandbranch[2]}")
```
Comment (ajithnn) [inline mygit.rb:154]: Spacing should be 2 spaces per tab. Please adjust your IDE settings.

File: mygit.rb [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+        $LOG.debug("<MyGit> <mygit_create> : Checking for git exists : #{GitExist?()}")
+        if !GitExist?()
+                $LOG.debug("<MyGit> <mygit_create> : Exiting because git doesn`t exist")
```
Comment (ajithnn) [inline mygit.rb:157]: Provide message on stdout  asking user to install git

File: mygit.rb [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+                exit
+        end
+        repo = repoandbranch[1]
+        branch =  repoandbranch[2]
+        $LOG.debug("<MyGit> <mygit_create> : Starting to clone the repo #{repo}")
+        system "git clone #{repo}"
```
Comment (ajithnn) [inline mygit.rb:163]: Extract git command output and add to debug logs.

File: mygit.rb [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+        writetofile("pullreq.txt", create_pull.number, repo)
+        writetofile("reviewreq.txt", reqreview.id , repo)                        
+  
+                                
+end
+
+def mygit_status()
+        client = octoKitInit()
+        repo = getreponame()
+        repo_directory = File.open("repo.txt").read
```
Comment (ajithnn) [inline mygit.rb:332]: All file opens should also close the file. Else handle operations in a block so ruby closes the file at the end of the block.

File: mygit.rb [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+
+                                                                                                                
+                puts status.state
+
+        end
+
+                
+        if status.state == 'APPROVED' 
+                puts "<MyGit> : your review is approved"
+                system "git add ."
```
Comment (ajithnn) [inline mygit.rb:356]: git add . - adds all files in the folder > subfolder > ** .
You need to ask for .gitignore file in init stage. Add entries to it.

File: mygit.rb [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+                system "git commit -m 'from Girish's App'"
```
Comment (ajithnn) [inline mygit.rb:357]: Commit message should be a param from the user.

Must allow multiline commit message.

File: mygit.rb [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+                system "git push -u origin #{directory}"
+        end
+        $LOG.debug("<MyGit> <mygit_prepare> : git pushed #{directory}")
+end
+
+def mygit_request(reviewers)
+        $LOG.debug("Request Called")
+        $LOG.debug("<MYGIT> <mygit_request>module : Check if git exists? : #{GitExist?()}")
+        reviewers.shift
+        repo_directory = File.open("repo.txt")
```
Comment (ajithnn) [inline mygit.rb:283]: Close

File: mygit.rb [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+        Dir.chdir(repo)
+        if base.nil?
+        $LOG.debug("<MyGit> <mygit_prepare> : Reading base from file as it is not provided in function parameter")
+        base = readhidden(repo, "Base.txt")
+        end
+
+        writetofile("Base.txt",base,repo)
+        
+        $LOG.debug("<MyGit> <mygit_prepare> : Calling rebase")
+        system "git rebase #{base}"
```
Comment (ajithnn) [inline mygit.rb:267]: What happens when rebase turns up merge conflicts , test and handle it.

File: mygit.rb [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+
+def readhidden(repo,file)
+
+        Dir.chdir ".#{repo}"
+        if !File.exist?(file)
+                $LOG.debug("<MyGit> <readhidden> : #{file} doesn't exist")
+                puts " <MyGit> : Error : File doesn`t exist"
+                exit
+        end
+        data = File.open(file, 'r').read
```
Comment (ajithnn) [inline mygit.rb:241]: Close

File: mygit.rb [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+        Dir.mkdir(lastmod)
+        `mv #{lastmod} .#{lastmod}`
+        $LOG.debug("<MyGit> <mygit_create> : Created the Hidden Directory to store data")
+        system "git init"
+        system "git pull"
+
+        if !(branch==repo)
+                system "git branch #{branch}"
+                system "git checkout #{branch}"
+                writetofile("Direx.txt",branch,repo)
```
Comment (ajithnn) [inline mygit.rb:191]: File names like repo.txt / Direx.txt / gitconfig.txt , must be constants , If the filename needs to be changed then only the constants value needs to be edited.

File: mygit.rb [hunk 20]
```diff
@@ -0,0 +1,365 @@
...
+                                        
+        Dir.chdir "#{lastmod}"
+        username = repo.split('/')[3]
+        repo=(repo.split('/')[4]).split('.')[0]
+        puts repo
+        repolink  = "#{username}"+"/"+"#{lastmod}"
+        Dir.mkdir(lastmod)
+        `mv #{lastmod} .#{lastmod}`
+        $LOG.debug("<MyGit> <mygit_create> : Created the Hidden Directory to store data")
+        system "git init"
```
Comment (ajithnn) [inline mygit.rb:185]: Does this handle error cases from git command. When git pull is run and internet connection does not exist what happens ?

**Review by ajithnn — COMMENTED** [review]
Please handle these

File: mygit.rb (hunk 7 above)
Comment (GirishPMallya) [reply mygit.rb:81]: I am using it to navigate between the git directory and the hidden directory I have created. The hidden directory contains all the files relevant to the application, like the git config, base branch etc..

File: mygit.rb (hunk 9 above)
Comment (GirishPMallya) [reply mygit.rb]: > For line by line reading of file

would this be necessary even if my file contains only a single line? as is the case with my file, which contains only the repo directory.

**Pull request closed without merging**